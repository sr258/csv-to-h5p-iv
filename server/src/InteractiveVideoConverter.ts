import papa from 'papaparse';
import fs from 'fs/promises';

import Converter from './Converter';
import InteractiveVideo from './content-types/InteractiveVideo';

export default class InteractiveVideoConverter extends Converter {
    public static create = async (): Promise<InteractiveVideoConverter> => {
        const converter = new InteractiveVideoConverter();
        await converter.init();
        return converter;
    };

    public parse = async (csvFile: string) => {
        let unparsedCsv: string;
        try {
            unparsedCsv = await fs.readFile(csvFile, {
                encoding: 'utf-8'
            });
        } catch (error) {
            throw new Error(
                `There was an error while reading the file ${csvFile}: ${error.message}`
            );
        }
        const parsedCsv = papa.parse(unparsedCsv, {
            header: false,
            delimiter: ';',
            skipEmptyLines: true
        });

        if (parsedCsv.meta.aborted) {
            throw new Error('The CSV could not be parsed');
        }

        if (parsedCsv.errors && parsedCsv.errors.length > 0) {
            throw new Error(
                `There were errors while parsing the CSV: ${parsedCsv.errors
                    .map((e) => `Row ${e.row} ${e.type} ${e.code} ${e.message}`)
                    .join(' ')}`
            );
        }

        // Filter empty rows
        const data = parsedCsv.data.filter((row: string[]) =>
            row.some((cell) => cell != '')
        );

        return InteractiveVideo.fromCsvRows(data as string[][]);
    };
}
