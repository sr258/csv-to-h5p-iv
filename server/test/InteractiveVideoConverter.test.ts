import path from 'path';
import fs from 'fs-extra';
import { withFile } from 'tmp-promise';

import User from '../src/User';
import InteractiveVideoConverter from '../src/InteractiveVideoConverter';

describe('Interactive Video Converter', () => {
    it('initializes and can parse csv', async () => {
        // prepare
        const converter = await InteractiveVideoConverter.create();

        // test
        const result = await converter.parse(
            path.join(__dirname, './__fixtures__/file2.csv')
        );
        expect(result).toMatchSnapshot();
    });
});
