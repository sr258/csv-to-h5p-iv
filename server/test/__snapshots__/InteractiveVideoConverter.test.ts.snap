// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interactive Video Converter initializes and can parse csv 1`] = `
InteractiveVideo {
  "interactions": Array [
    DragText {
      "correctText": undefined,
      "correctTime": undefined,
      "end": 24.882,
      "incorrectText": "Try again",
      "incorrectTime": undefined,
      "splitAdaptiveAction": [Function],
      "start": 22.882,
      "task": "首先， 我*起身 hei2 san1*
sau2 sin1, ngo5 ___. 
First, I ___.

然後， 我*換衫 wun6 saam1*
jin4 hau6, ngo5 ___. 
Then, I ___.

跟住， 我*食包 sik6 baau1*
gan1 zyu6, ngo5 ___. 
After that, I ___.

最後， 我*搭車 daap3 ce1*
zeoi3 hau6, ngo5 ___. 
Finally, I ___.",
      "taskDescription": "將單詞拖到正確嘅方格內。

zoeng1 daan1 ci4 to1 dou3 zing3 kok3 ge3 fong1 gaak3 noi6.

Drag the words into the correct boxes.",
      "title": "Ordering Events",
      "type": "DragText",
    },
    FreeTextQuestion {
      "correctText": undefined,
      "correctTime": undefined,
      "end": 31.87,
      "incorrectText": "Review slide 1",
      "incorrectTime": 29.87,
      "placeholder": "媽媽 maa4 maa1 Mom
爸爸 baa4 baa1 Dad
﻿婆婆 po4 po2 Grandma
﻿爺爺 je4 je2 Grandpa",
      "question": "邊個同你搭巴士?

bin1 go3 tung4 nei5 daap3 baa1 si2?

Who went on the bus with you?",
      "splitAdaptiveAction": [Function],
      "start": 29.87,
      "title": "Who went on the bus with you?",
      "type": "FreeTextQuestion",
    },
    TrueFalse {
      "answer": true,
      "correctText": undefined,
      "correctTime": undefined,
      "end": 49,
      "incorrectText": undefined,
      "incorrectTime": undefined,
      "question": "十分鐘之後有巴士.

sap6 fan1 zung1 zi1 hau6 jau5 baa1 si2.

There is a bus after 10 minutes.",
      "splitAdaptiveAction": [Function],
      "start": 47,
      "title": "10 Minutes",
      "type": "TrueFalse",
    },
    Blanks {
      "correctText": undefined,
      "correctTime": undefined,
      "end": 120.199,
      "incorrectText": undefined,
      "incorrectTime": undefined,
      "splitAdaptiveAction": [Function],
      "start": 118.199,
      "task": "我冇位坐，又*孭/me1*住個大書包，唔係咁開心。

ngo5 mou5 wei2 co5, jau6 __ zyu6 go3 daai6 syu1 baau1, m4 hai6 gam3 hoi1 sam1.

I don't have a seat to sit at, and I have to _____ a big backpack. I'm not very happy.",
      "taskDescription": "喺空格內填上適當嘅中文字或粵拼。

hai2 hung1 gaak3 noi6 tin4 soeng5 sik1 dong3 ge3 zung1 man4 zi6 waak6 jyut6 ping3

Fill in the missing word(s) with Chinese characters or Jyutping.",
      "title": "Carry",
      "type": "Blanks",
    },
    DragQuestion {
      "correctText": undefined,
      "correctTime": undefined,
      "end": 179.86,
      "entries": Array [
        Object {
          "dragableText": "叔叔

suk1 suk1

Uncle",
          "dropZoneCaption": "第一 dai6 jat1 First",
        },
        Object {
          "dragableText": "媽媽同我

maa4 maa1 tung4 ngo5

Mom and me",
          "dropZoneCaption": "第二 dai6 ji6 Second",
        },
        Object {
          "dragableText": "婆婆

po4 po2

Grandma",
          "dropZoneCaption": "第三 dai6 saam1 Third",
        },
      ],
      "incorrectText": undefined,
      "incorrectTime": undefined,
      "splitAdaptiveAction": [Function],
      "start": 177.86,
      "taskDescription": "拖放：佢哋以乜嘢次序上巴士？to1 fong3: keoi5 dei6 ji5 mat1 je5 ci3 zeoi6 soeng5 baa1 si2? Drag and Drop: In what order did these people get on the bus?",
      "type": "DragQuestion",
    },
    FreeTextQuestion {
      "correctText": undefined,
      "correctTime": undefined,
      "end": 255,
      "incorrectText": undefined,
      "incorrectTime": undefined,
      "placeholder": "十五分鐘; sap6 ng5 fan2 zung1; 15 minutes",
      "question": "由你屋企到學校需要幾長時間？jau4 nei5 uk1 kei5 dou3 hok6 haau6 seoi1 jiu3 gei2 coeng4 si4 gaan3? How long does it take to get from your house to school?",
      "splitAdaptiveAction": [Function],
      "start": 253,
      "title": "How long does it take to get from your house to school?",
      "type": "FreeTextQuestion",
    },
    MultipleChoice {
      "correctText": undefined,
      "correctTime": undefined,
      "end": 256,
      "incorrectText": undefined,
      "incorrectTime": undefined,
      "options": Array [
        Object {
          "correct": true,
          "text": "我 ngo5 Me",
        },
        Object {
          "correct": true,
          "text": "我 ngo5 Me",
        },
        Object {
          "correct": false,
          "text": "爸爸 baa4 baa1 Dad",
        },
        Object {
          "correct": false,
          "text": "朋友 pang4 jau5 Friend",
        },
        Object {
          "correct": true,
          "text": "叔叔 suk1 suk1 Uncle",
        },
        Object {
          "correct": false,
          "text": "姨姨 ji1 ji1 Aunt",
        },
        Object {
          "correct": false,
          "text": "同學 tung4 hok6 Schoolmate",
        },
        Object {
          "correct": true,
          "text": "婆婆 po4 po2 Grandma",
        },
        Object {
          "correct": false,
          "text": "爺爺 je4 je2 Grandpa",
        },
        Object {
          "correct": false,
          "text": "老師 lou5 si1 Teacher",
        },
      ],
      "question": "故事入面，有邊個搭巴士?

gu3 si6 jap6 min6, jau5 bin1 go3 daap3 baa1 si2?

Who took the bus in the tory?",
      "splitAdaptiveAction": [Function],
      "start": 254,
      "title": "Who took the bus?",
      "type": "MultipleChoice",
    },
    Text {
      "correctText": undefined,
      "correctTime": undefined,
      "end": 1,
      "incorrectText": undefined,
      "incorrectTime": undefined,
      "splitAdaptiveAction": [Function],
      "start": 1,
      "text": "This is some text that you can display without any interaction

You can give instructions, add copyright notices etc.",
      "type": "Text",
    },
  ],
  "mediaLink": "https://www.youtube.com/watch?v=FwdMr6cYcN0&amp;ab_channel=%E5%86%9A%E5%94%AA%E5%94%A5%E7%B2%B5%E6%96%87%E8%AE%80%E6%9C%ACHambaanglaangCantoneseGradedReaders",
  "title": "《搭巴士》Taking The Bus (Child, Lv 1): Interactive Video: Cantonese Version",
}
`;
